app/com/linkedin/drelephant/spark/heuristics/StagesAnalyzer.scala

```diff
@@ -0,0 +1,498 @@
+    *
+    * @param stageId stage ID
+    * @param median median data size for tasks.
+    * @param maximum maximum data size for tasks.
+    * @param description type of data.
+    * @param details information and recommendations -- any new recommendations
+    *                from analyzing the stage for data skew will be appended.
+    */
+  private def checkSkewedData(
+      stageId: Int,
+      median: Double,
+      maximum: Double,
+      description: String,
+      details: ArrayBuffer[String]): Severity = {
+    val severity = taskSkewThresholds.severityOf(maximum / median)
+    if (hasSignificantSeverity(severity)) {
+      details += s"Stage $stageId has skew in $description (median is " +
+        s"${MemoryFormatUtils.bytesToString(median.toLong)}, " +
+        s"max is ${MemoryFormatUtils.bytesToString(maximum.toLong)})."
+    }
+    severity
+  }
+
+  /**
+    * Check the stage for long running tasks.
+    *
+    * @param stageId stage ID.
+    * @param stageData stage data.
+    * @param medianTime median task run time.
+    * @param curNumPartitions number of partitions for the Spark application
+    *                         (spark.sql.shuffle.partitions).
+    * @return results of long running task analysis for the stage
+    */
+  private def checkForLongTasks(
+      stageId: Int,
+      stageData: StageData,
+      medianTime: Option[Double],
+      curNumPartitions: Int): SimpleStageAnalysisResult = {
+    val longTaskSeverity = stageData.taskSummary.map { distributions =>
+      taskDurationThresholds.severityOf(distributions.executorRunTime(DISTRIBUTION_MEDIAN_IDX))
+    }.getOrElse(Severity.NONE)
+    val details = new ArrayBuffer[String]
+    if (hasSignificantSeverity(longTaskSeverity)) {
+      val runTime = Utils.getDuration(medianTime.map(_.toLong).getOrElse(0L))
+      val maxData = Seq(stageData.inputBytes, stageData.shuffleReadBytes, stageData.shuffleWriteBytes,
+        stageData.outputBytes).max
+      details += s"Stage $stageId: median task run time is $runTime."
+      if (stageData.numTasks >= maxRecommendedPartitions) {
+        if (maxData >= maxDataProcessedThreshold) {
+          val inputBytes = MemoryFormatUtils.bytesToString(stageData.inputBytes)
+          val shuffleReadBytes = MemoryFormatUtils.bytesToString(stageData.shuffleReadBytes)
+          val shuffleWriteBytes = MemoryFormatUtils.bytesToString(stageData.shuffleWriteBytes)
+          details += s"Stage $stageId has $inputBytes input, $shuffleReadBytes shuffle read, " +
+            "$shuffleWriteBytes shuffle write. Please try to reduce the amount of data being processed."
+        } else {
+          details += s"Stage $stageId: please optimize the code to improve performance."
+        }
+      }
+      else {
+        if (stageData.inputBytes > 0) {
+          // The stage is reading input data, try to increase the number of readers
+          details += s"Stage $stageId: please set DaliSpark.SPLIT_SIZE to a smaller " +
```

Comment:
I think this is Dali specific suggestion . What if user is not using Dali or for open source , should we give more general suggestion regarding split size ? .

Reply:
Yes it is Dali and LinkedIn specific, so not appropriate for open source. I can modify the suggestion to make it more general, but it would be good to have the more specific suggestion for our users. What would be the best way to customize? Perhaps there could be a heuristic configuration, with a map of issue type to recommendation?

Reply:
I think , we can enhance the details and be specific to LinkedIn user and other outside users . Something like "please set DaliSpark.SPLIT_SIZE to a smaller : Linkedin User ,
                            please set mapreduce.input.fileinputformat.split.maxsize : Outside user" .

I think , it make sense to have map of issue type to recommendation